model/src/main/kotlin/Hash.kt

```diff
@@ -19,68 +19,159 @@
-        require(algorithm in VERIFIABLE) {
-            "Cannot verify algorithm '$algorithm'. Supported algorithms are $VERIFIABLE."
+        require(canCalculate) {
+            "Cannot calculate values for '${javaClass.simpleName}' hash."
+        }
+
+        return file.hash(names.first()) == value
+    }
+
+    /**
+     * The Message-Digest 5 hash algorithm that was used to calculate the [value],
+     * see [MD5](https://en.wikipedia.org/wiki/MD5).
+     */
+    class Md5(value: String) : Hash(listOf("MD5"), Regex("[0-9A-Fa-f]{32}"), value)
+
+    /**
+     * The Secure Hash Algorithm 1 that was used to calculate the [value],
+     * see [SHA-1](https://en.wikipedia.org/wiki/SHA-1).
+     */
+    class Sha1(value: String) : Hash(listOf("SHA-1", "SHA1"), Regex("[0-9A-Fa-f]{40}"), value)
+
+    /**
+     * The Secure Hash Algorithm 2 with 256 bits that was used to calculate the [value],
+     * see [SHA-256](https://en.wikipedia.org/wiki/SHA-256).
+     */
+    class Sha256(value: String) : Hash(listOf("SHA-256", "SHA256"), Regex("[0-9A-Fa-f]{64}"), value)
+
+    /**
+     * The Secure Hash Algorithm 2 with 384 bits that was used to calculate the [value],
+     * see [SHA-384](https://en.wikipedia.org/wiki/SHA-384).
+     */
+    class Sha384(value: String) : Hash(listOf("SHA-384", "SHA384"), Regex("[0-9A-Fa-f]{96}"), value)
+
+    /**
+     * The Secure Hash Algorithm 2 with 512 bits that was used to calculate the [value],
+     * see [SHA-512](https://en.wikipedia.org/wiki/SHA-512).
+     */
+    class Sha512(value: String) : Hash(listOf("SHA-512", "SHA512"), Regex("[0-9A-Fa-f]{128}"), value)
+
+    /**
+     * An unknown hash algorithm that was used to calculate the [value].
+     */
+    class Unknown(value: String) : Hash(listOf("UNKNOWN"), Regex("\\S.+\\S"), value, false)
+
+    /**
+     * No hash at all, both [names] and [value] are empty.
+     */
+    class None : Hash(listOf(""), Regex(""), "", false)
+}
+
+class HashDeserializer : StdDeserializer<Hash>(Hash::class.java) {
+    override fun deserialize(p: JsonParser, ctxt: DeserializationContext): Hash {
+        val node = p.codec.readTree<JsonNode>(p)
+        return if (node.isTextual) {
+            val hashValue = node.textValue()
+            if (p.nextFieldName() == "hash_algorithm") {
+                val hashAlgorithm = p.nextTextValue()
+                Hash.create(hashValue, hashAlgorithm)
+            } else {
+                Hash.create(hashValue)
+            }
+        } else {
+            Hash.create(node["value"].textValue(), node["algorithm"].textValue())
         }
+    }
+}
 
-        return file.hash(algorithm.toString()) == value
+class HashSerializer : StdSerializer<Hash>(Hash::class.java) {
+    override fun serialize(hash: Hash, gen: JsonGenerator, provider: SerializerProvider) {
+        gen.writeStartObject()
+        gen.writeStringField("value", hash.value)
+        gen.writeStringField("algorithm", hash.names.first())
+        gen.writeEndObject()
```

Comment:
Wouldn't adding `@JsonIgnore` to `names` and `valueRegex` achieve the same without requiring a custom serializer?

Reply:
Not quite... from `names` the first value actually *should* get serialized. And `canCalculate` would need to get ignored, too.